### ISSUE TITLE
XML report merging doesn't work without renaming reports to `detekt.xml` 

### ISSUE THREAD

**Issue by pbreault** [detekt/detekt#4192]
## Expected Behavior
[XML Report merging](https://detekt.github.io/detekt/reporting.html#merging-reports) should work out of the box when following the examples in the documentation. 

## Observed Behavior
Report merging fails silently for android modules ( and maybe kotlin libraries? )  because it expects all xml reports to be named `detekt.xml`, while by default they are named   `debug.xml`, `main.xml`, etc.

The name seems to be hardcoded in:  `io.gitlab.arturbosch.detekt.Detekt#getTargetFileProvider`

## Steps to Reproduce

1)  Unzip the attached [Repro Project](https://github.com/detekt/detekt/files/7365640/detekt-report-merge-repro-project.zip)
2)  run `./repro.sh`
3) Notice that there is nothing under `$rootProject/build/reports/detekt`

## Your Environment
* Version of detekt used: 1.18.1
* Version of Gradle used (if applicable): 7.2
* Operating System and version: MacOS 11.5.1

**Comment by cortinico** [comment]
>     1. Unzip the attached [Repro Project](https://github.com/detekt/detekt/files/7365640/detekt-report-merge-repro-project.zip)

The link is broken so I can't really say what's going on.

How are you configuring the reports? Can I ask you to try the latest `-SNAPSHOT` and follow the documentation suggested in this PR? https://github.com/detekt/detekt/pull/4064
In the upcoming version of Detekt we fixed how reports are configured therefore this issue might be already solved.

**Comment by 3flex** [comment]
This was closed in error... It's not yet fixed, only a test related to it was fixed.

**Comment by RSBat** [comment]
I wanted to see if this is fixable without switching to cross-project publication and here's what I found:

Unless it is intentional, there is a problem with implementation of `getTargetFileProvider`. It returns `RegularFileProperty` that has fixed value, so it is not updated when report's `outputLocation` changes. This explains why the order of the calls mentioned before matters. 

Correct implementation would return property that tracks report's `outputLocation`, but that is not easy. If (for simplicity) we replace method body with just `return report.outputLocation` we would get the following error when running gradle:
```
* What went wrong:
Could not determine the dependencies of task ':detektReportMergeSarif'.
> Property 'outputLocation' is declared as an output property of DetektReport(type='SARIF', required=property 'required', outputLocation=property 'outputLocation') (type DetektReport) but does not have a task associated with it.
```

This is a [known problem in gradle](https://github.com/gradle/gradle/issues/6619) and it is partially fixed, but I still could not get the code to work. However I found a workaround:
```kotlin
private fun getTargetFileProvider(
        report: DetektReport
    ): RegularFileProperty {
        val provider = objects.fileProperty()
        val temp = objects.fileProperty()
	temp.set(File("")) // set value, so `map` will be executed
	provider.fileProvider(temp.map {
            val isEnabled = report.required.getOrElse(DetektExtension.DEFAULT_REPORT_ENABLED_VALUE)
            if (isEnabled) {
	        report.outputLocation.asFile.orNull ?: reportsDir.getOrElse(defaultReportsDir.asFile)
                    .resolve("${DetektReport.DEFAULT_FILENAME}.${report.type.extension}")
            } else {
                null
	    }
	})
        return provider
    }
```
Tested it with my Android project and it works fine.


Of course after I did all of that I found an even simpler workaround: add `def reportMerge = ` to the start of the line that registers merge task. It changes `reportMerge.configure` to be a call on task provider instead of task, thus delaying configuration enough for the correct directory to be set.

**Comment by RSBat** [comment]
> I haven't had time to make changes to docs or plugin but the real issue here is that there's a way to reliably share artifacts between Gradle projects, similar to how we do it within the detekt project for JaCoCo, and it's not being recommended in the detekt docs for XML report merging.
> 
> See here for more: https://docs.gradle.org/7.2/userguide/cross_project_publications.html

I tried to set up cross project artifact sharing and almost got it to work in this project: https://gist.github.com/RSBat/5c14485aec4fdf8a03173aeffd0f978a

However I can't get it to merge reports when any of `detekt` tasks fail. I know it is possible because jacoco plugin does this even when tests fail. I've tried to check jacoco's implementation, but I couldn't find anything that would explain the difference in behaviour.

@3flex since you mentioned it first, maybe you have some ideas?

**Comment by 3flex** [comment]
I know this is an issue however I haven't looked at it yet, but may do so at some point when time and motivation allows.

We're of course open to PRs.

**Comment by RSBat** [comment]
Figured it out! The reason jacoco plugin works is that test tasks throw `VerificationException` on failure. This allows aggregation task to still run even though test tasks failed (but only if `--continue` is used). Changing detekt plugin to throw the same exception combined with the [config I posted before](https://gist.github.com/RSBat/5c14485aec4fdf8a03173aeffd0f978a) makes everything work. Unfortunately `VerificationException` is an incubating feature added in 7.4 and is not documented anywhere. The only info I could find was the PR that added it https://github.com/gradle/gradle/pull/18837.

To sum up:
- The issue is caused by #4856 - report location properties are not actually lazy and this creates dependencies on configuration order.
- The only workaround that worked for me was to rename reports to `detekt.sarif`
- For report aggregation via artifact sharing to work properly detekt tasks have to throw `VerificationException` on failure. This would require bumping minimum gradle version to 7.4

**Comment by namnh-0652** [comment]
> I tried to set up cross project artifact sharing and almost got it to work in this project: ?> https://gist.github.com/RSBat/5c14485aec4fdf8a03173aeffd0f978a

I converted to `groovy` version but it did not work on my multiple modules `android` project. Even the command `./gradlew detekt reportMerge --continue` was run successfully, but no output file `merge.sarif` found.
```groovy
buildscript {
    repositories {
        mavenCentral()
    }
    dependencies {
        classpath "io.gitlab.arturbosch.detekt:detekt-gradle-plugin:1.22.0"
    }
}

import io.gitlab.arturbosch.detekt.Detekt
import io.gitlab.arturbosch.detekt.report.ReportMergeTask

configurations {
    create("reportMergeInputs") {
        setCanBeResolved(true)
        setCanBeConsumed(false)
        setVisible(false)

        attributes.attribute(Category.CATEGORY_ATTRIBUTE, objects.named(Category, Category.VERIFICATION))
        attributes.attribute(VerificationType.VERIFICATION_TYPE_ATTRIBUTE, objects.named(VerificationType, "detekt-report"))
    }
}

tasks.register("reportMerge", ReportMergeTask) {
    input.from(configurations.getByName("reportMergeInputs").incoming.artifacts.artifactFiles)
    output = rootProject.layout.buildDirectory.file("reports/detekt/merge.sarif")
}

allprojects {
    apply plugin: "io.gitlab.arturbosch.detekt"
    detekt {
        config = files("$rootDir/config/detekt/detekt.yml")
    }
    def subproject = this
    def configurationName = subproject.name + "DetektOutput"
    subproject.configurations {
        create(configurationName) {
            println(subproject.name)
            println(configurationName)
            setCanBeResolved(false)
            setCanBeConsumed(true)
            setVisible(false)

            attributes.attribute(Category.CATEGORY_ATTRIBUTE, objects.named(Category, Category.VERIFICATION))
            attributes.attribute(VerificationType.VERIFICATION_TYPE_ATTRIBUTE, objects.named(VerificationType, "detekt-report"))
        }
    }
    tasks.withType(Detekt) { a ->
        jvmTarget = "1.8"
        reports {
            xml.required.set(true)
            html.required.set(true)
            txt.required.set(true)
            sarif.required.set(true)
            md.required.set(true)
        }
        basePath = rootProject.projectDir.absolutePath
        subproject.afterEvaluate {
            subproject.configurations.getByName(configurationName).outgoing {
                artifact(task.sarifReportFile) {
                    setType(ArtifactTypeDefinition.BINARY_DATA_TYPE)
                    builtBy(task) // task will be executed if we need its report
                }
            }
        }
    }
}
```

**Comment by 3flex** [comment]
Please review #6964 which should fix this.

@tabasavr thank you for your investigation, the proposed PR is taking advantage of much of the work you did.
